### ISSUE TITLE
Failure Threshold revert to normal non auto-renew contribution/membership

### ISSUE THREAD

**Issue by shaneonabike** [iATSPayments/com.iatspayments.civicrm#462]
## Overview

So I have an organization I work with that has a need that I think is a possible enhancement, but I think I need some guidance on how to go about doing it.

Essentially, the idea is that instead of simply pushing the contribution to the next year (after X failures) we actually cancel the auto-renewal of their contribution. For memberships this is pretty ideal, because in most organizations that I work for this means access to specific privileges and other things. So when people login they clearly see that they need to renew.

## Present situation

1. User sets up auto-renew on membership
2. Their CC expires and they get a new one (forgetting to update their account)
3. The payment goes through X times and fails
4. The membership auto-renewal gets bumped to the following year to retry
5. Workers need to fix memberships which are bumped up, contact members, and reprocess everything

## Suggested solution

+ Provide a means that would allow the cancellation of the auto-renewal contribution after the threshold instead of pushing it up to the next year as it does
+ Document that it is advised to send emails of failed transactions to their members so that they are aware of what is happening

Thoughts?

cc @adixon

**Comment by adixon** [comment]
I think there's already a feature to send emails of failed transactions to the cardholder (or at least a discussion and/or PR, it sounds familiar).

Re: auto cancellation, I'm not clear about the problem you're trying to solve here.

Agreed that the current behaviour of 3 attempts and then pushing to the next date was designed as a fallback behaviour for contributions, not memberships. Maybe an alternative is provide an option to "suspend" the recurring contribution after three failures? Or is that what you're proposing? Actually cancelling seems like the wrong thing to do, that feels like something that should be done by an actual person to confirm it.

**Comment by adixon** [comment]
Take a look at the UI configuration option "Email receipt for a Contribution if Recurring payment fails - with error message", I think that would be what you want to do to alert a custom that their cc has failed.

**Comment by shaneonabike** [comment]
> I think there's already a feature to send emails of failed transactions to the cardholder (or at least a discussion and/or PR, it sounds familiar).

You bet it does! I was mainly saying that we should document somewhere that this is done if we implement my solution.

> Auto cancellation

I probably used the wrong terminology here. I meant that we essentially stop attempting to process the contribution and don't push the next date to the next year. So a suspend is a great idea. I would imagine that in that case we could do a few things

+ Display a message on the appropriate Contribution page to indicate why it was suspended
+ Allow a person to recomplete the Contribution page with their new details

**Comment by shaneonabike** [comment]
.. and just to be totally clear I think you can leave how it works presently because maybe people would like that, but I am proposing to add a new option that says something like

```After the failure threshold has been reached suspend the contribution from further being processed````

and then in the help text it would read

``` It is recommended that you send a Email receipted to the people who have setup auto-renewal upon failure as this option will simply stop processing after the threshold is reached.```

Hopefully, that's a bit clearer what I'm trying to get at... and I don't know if you think it's easy and/or logical

**Comment by adixon** [comment]
Yup, PR welcome to provide a new configuration re: behaviour after threshold failures.

If you're doing annual auto-renewals of membership AND recurring monthly donations, it might make sense to have different settings, but that would also be complicated, so let's start with a single setting that applies across the board.

**Comment by shaneonabike** [comment]
Happy to do this. Can you give me some pointers though on the suspend portion. I see you check for the threshold (here)[https://github.com/iATSPayments/com.iatspayments.civicrm/blob/5dcc8715ac316e5c8cf48d9c2598e36b45d34cdc/CRM/Iats/Transaction.php#L119] 

But I'm a bit lost as to what is happening after that (if it fails say after 3). I see you are creating a ```ContributionRecur``` if we have a failure. I guess I can assume that basically when it fails the _existing one_ never gets processed again so it creates a new ContributionRecur to process the next one right?

In our case, if the threshold was reached I could technically update the existing to _Suspend_ ?

cc @adixon

**Comment by adixon** [comment]
No, I think you're misunderstanding the code. The 'creation' of a ContributionRecur is actually just updating the existing one, I'm not creating any new schedules.

The behaviour if it's over the threshold is defined by this comment:

// otherwise, we'll keep the next collection date as passed in from the caller

So instead we need to test your new configuration switch, and if we're only suspending, then we also push the next contribution back and set the status to 'pending'. 

The thing to know at this stage of the code is that, for "bulletproof" reasons, the recurring contribution has already been updated on the assumption that the recurring contribution succeeded (to avoid recharging a schedule where the payment goes through but that information doesn't get recorded in civi, for a variety of possible reasons). So we're updating that recurring schedule to handle cases where that's not true.

**Comment by shaneonabike** [comment]
@adixon I'm finally working on this. I'm curious why we need to set this to `Pending` rather than `Cancelled`. I feel that Pending might be confusing for people in the backend. Also, it isn't like you can modify the transaction to use another Credit Card (if you wanted to repair the problem b/c of changed credit card or something).

It's _totally_ possible there is a reason for this so that's why I ask.

Also, do you have any advice in testing this without using live transactions?

**Comment by shaneonabike** [comment]
@adixon nevermind on the testing. I just somewhat hacked the scheduled job to fail since iATS Test always approves transactions. So I was able to test this and it's working from what I can tell. Please feel free to give me feedback and any changes you think are appropriate in the merge request.

Maybe see you in San Fran!

**Comment by KarinG** [comment]
Hi Shane - couple of notes!

iATS Test -> 4111111111111111 - Allows you to test both Authorizations, Charges, and Rejections based on specific amounts listed on: https://appfrontier.com/iats-payments-salesforce-integration

I don't like removing data in general -> specifically removing the contribution_recur_id from a contribution record removes history and understanding.

I don't agree with auto-updating recurring contribution status to Cancelled. Auto-Cancellation would have a massive impact in how orgs with thousands of recurring donations every month currently handle those that fail for one reason or another. And different REJ reasons require different actions. 

In general we've avoided adding business logic. For example -> should status=Cancelled then also mean the Membership (if any) should be status=Cancelled?  For some orgs - maybe yes. For some - maybe no. And what if their Membership status gave them Event discounts -> should there perhaps we Cancelations for Events - for some orgs.

My thinking is for you to add this PR this Alan's Contribution Recur extension instead of to the core payment processing functionality. https://github.com/adixon/ca.civicrm.contributionrecur

-- KarinG

**Comment by shaneonabike** [comment]
I have to be honest I was pretty disheartened at your response @KarinG . I chatted quite some time with Alan, wrote in the ticket, and made sure that this is something that would work and would be approved. 

@KarinG if I can give some advice, _please_ chime in early so that other contributors don't spend the time trying to add a patch to find out that it's not something you (as maintainers) would be willing to add. It cost my organization time/money, and I felt proud to have contributed something I thought could (at the very least) be modified and integrated to add some nice additional features/alternatives to what existed. 

> 4111111111111111

Ahh yes but it won't fail afterwards which was the point of my original description at the top of this issue reported. I need to simulate a scenario where CC are in _auto-renew_ and fail on thresholds reach on the next years attempt at renewals not the initial attempt :)

> I don't like removing data in general -> specifically removing the contribution_recur_id from a contribution record removes history and understanding.

That's fine it makes sense thanks for the tip.

> I don't agree with auto-updating recurring contribution status to Cancelled. Auto-Cancellation would have a massive impact in how orgs with thousands of recurring donations every month currently handle those that fail for one reason or another. And different REJ reasons require different actions.

I guess I could understand not setting it to `Cancelled` but it does seem sensible to set it to `Failed` or some other relevant status. Presently, after X failures _regardless of the rejection reason_ you skip to next year. That could also be deemed confusing for most orgs that I work with. If they were relying on that then the X is reached they don't get their funds, and don't understand there was an issue if the donator doesn't take action with the email they receive.

Furthermore, you aren't reacting to different REJ reasons from what I can gather. If you were then that would probably make sense. You have the threshold and no matter what REJ is happening after X it bumps to the next year. In lies the issue I was attempting to tackle.

> No business logic

Your perspective makes sense, but the current behaviour of pushing it to another year has the potential of confusing orgs as well. Most of the ones that I work with don't want that action especially when their Memberships are actually provide additional services and support (and they rely on those funds as well).

> Alan's recurring extension

I checked this out and it's not that appropriate to the issue I am addressing. It handles a totally different scenario then what I want.
